crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,289 @@
+                let remainder = remainder.to_string().trim_end_matches('0').to_owned();
+                BalanceVariant::Denominated(format!(
+                    "{}.{}{}{}",
+                    units, remainder, symbol, token_metadata.symbol
+                ))
+            } else {
+                BalanceVariant::Denominated(format!(
+                    "{}{}{}",
+                    units, symbol, token_metadata.symbol
+                ))
+            }
+        } else {
+            BalanceVariant::Default(n)
+        }
+    }
+}
+
+impl Display for BalanceVariant {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            BalanceVariant::Default(balance) => f.write_str(&balance.to_string()),
+            BalanceVariant::Denominated(input) => f.write_str(input),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cmd::extrinsics::parse_balance;
+
+    use super::*;
+
+    #[test]
+    fn correct_balances_parses_success() {
+        assert!(
+            parse_balance("500DOT").is_ok(),
+            "<500DOT> was not parsed correctly"
+        );
+        assert!(
+            parse_balance("500").is_ok(),
+            "<500> was not parsed correctly"
+        );
+    }
+
+    #[test]
+    fn incorrect_balances() {
+        assert!(
+            parse_balance("500%").is_err(),
+            "expected to fail parsing incorrect balance"
+        );
+    }
+
+    #[test]
+    fn balance_variant_denominated_success() {
+        let tm = TokenMetadata {
+            denomination: 12,
+            symbol: String::from("DOT"),
+        };
+        let bv = parse_balance("500MDOT").expect("successful parsing. qed");
+        assert!(
+            bv.denominate_balance(&tm).is_ok(),
+            "balances could not be denominated correctly"
+        );
+    }
+
+    #[test]
+    fn balance_variant_denominated_incorrect_token() {
+        let tm = TokenMetadata {
+            denomination: 12,
+            symbol: String::from("DOT"),
+        };
+        let bv = parse_balance("500MKSM").expect("successful parsing. qed");
+        assert!(
+            bv.denominate_balance(&tm).is_err(),
+            "balances denominated should fail because of an incorrect token"
+        );
+    }
+
+    #[test]
+    fn balance_variant_denominated_equal() {
+        let denomination: u128 = format!("1{}", "0".repeat(12)).parse().unwrap();
+        let tm = TokenMetadata {
+            denomination,
+            symbol: String::from("DOT"),
+        };
+        let balance: Balance = 500 * 1_000_000 * denomination;
+        let bv = parse_balance("500MDOT").expect("successful parsing. qed");
+        let balance_parsed = bv.denominate_balance(&tm).expect("successful parsing. qed");
+        assert_eq!(balance, balance_parsed);
+    }
+
+    #[test]
+    fn balance_variant_denominated_equal_fraction() {
+        let denomination: u128 = format!("1{}", "0".repeat(12)).parse().unwrap();
```

Comment:
All tests for the `denominate_balance` are assuming `12` decimals. There should be some other values for this, there are some bugs this will show.